### DISCUSSION TITLE
Possibility of increasing time taken to execute gnark object creation functions

### DISCUSSION THREAD

**Discussion by namrapatel** [Consensys/gnark#778] [Q&A]
Hello!

I'm working on a cloud function that generates proofs for circuits written using gnark. The reason I'm using a cloud function is because I want everything to be exceedingly fast and I've narrowed down my bottleneck to the functions that are implemented for creating gnark objects from a buffer. The code below explains the problem, I'm wondering if you know of any way I can create the same gnark objects in a speedier manner?

```go
// Read constraint system from disk
err = readFromDisk(getZkFileName(request.CircuitType, request.CircuitArtifactUUID, "cs"), &buf)
if err != nil {
	log.Error().Err(err).Msg("Failed to decode cs from base64")
	return nil, err
}

// Build new constraint system object
cs := groth16.NewCS(ecc.BN254)
cs.ReadFrom(buf.Bytes()) // THIS LINE TAKES ~1000ms TO EXECUTE
buf.Reset()

// Read pk from disk
err = readFromDisk(getZkFileName(request.CircuitType, request.CircuitArtifactUUID, "pk"), &buf)
if err != nil {
	log.Error().Err(err).Msg("Failed to decode pk from base64")
	return nil, err
}

// Build new pk object
pk := groth16.NewProvingKey(ecc.BN254)
pk.ReadFrom(buf.Bytes())  // THIS LINE TAKES ~900ms TO EXECUTE
buf.Reset()
```

I've flagged the two lines that are taking around ~1000ms to execute and would love to hear if anyone has any ideas to make them faster. I've tried using the `msgpack.Unmarshal` from the msgpack library to try and build the same Constraint System and Proving Key objects, but then my proof fails to verify saying that "pairings do not match".

I'm happy to answer more questions if there are any clarifications I can make.

**Comment by ivokub** [comment]
Have you tried using `UnsafeReadFrom` instead of `ReadFrom`? It omits subgroup checks etc. If the proving key comes from a pre-trusted source, then it should be safe.
